src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -168,13 +169,89 @@ private void OnEventSourceChanged(object? _1, TaggerEventArgs _2)
+
+            if (frozenPartialSemantics)
+            {
+                // If we were asking for frozen partial semantics, then just proceed as normal, getting those tags
+                // quickly, but inaccurately.
+                await RecomputeTagsAsync(highPriority, frozenPartialSemantics: true, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                if (!_dataSource.SupportsFrozenPartialSemantics)
+                {
+                    // We're asking for normal expensive full tags, and this tagger doesn't support frozen partial
+                    // tagging anyways, so just proceed as normal, asking for expensive tags.
+                    await RecomputeTagsAsync(highPriority, frozenPartialSemantics: false, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // We're asking for expensive tags, and this tagger supports frozen partial tags.  Kick off the work
+                    // to do this expensive tagging, but attach ourselves to the requested cancellation token so this
+                    // expensive work can be canceled if new requests for frozen partial work come in.
+
+                    // We must have at least one request asking for full semantics.
+                    Contract.ThrowIfFalse(changes.Any(c => !c.FrozenPartialSemantics));
+
+                    // All those requests should have cancellation tokens provided with them.
+                    Contract.ThrowIfFalse(changes.Where(c => !c.FrozenPartialSemantics).All(c => c.NonFrozenComputationToken != null));
+
+                    // Get the first non-cancelled token if present.
+                    var nonFrozenComputationToken = changes.FirstOrNull(t => t.NonFrozenComputationToken?.IsCancellationRequested is false)?.NonFrozenComputationToken;
+
+                    // If there is no non-frozen cancellation token that has not been triggered yet, then all non-frozen work
+                    // has been canceled, and we can immediately bail out.
+                    if (nonFrozenComputationToken is null)
+                        return;
+
+                    // Otherwise, link this token with the main queue cancellation token and do the actual tagging.
+                    using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, nonFrozenComputationToken.Value);
+
+                    // Need a dedicated try/catch here since we're operating on a different token than the queue's token.
+                    try
+                    {
+                        await RecomputeTagsAsync(highPriority, frozenPartialSemantics, linkedTokenSource.Token).ConfigureAwait(false);
+                    }
+                    catch (Exception ex) when (FatalError.ReportAndPropagateUnlessCanceled(ex, linkedTokenSource.Token))
+                    {
+                    }
```

Comment:
This should probably catch/hide the exception from the caller since we're validating against a different cancellation token.

```suggestion
                    catch (Exception ex) when (FatalError.ReportAndPropagateUnlessCanceled(ex, linkedTokenSource.Token))
                    {
                    }
                    catch (Exception ex) when (ExceptionUtilities.IsCurrentOperationBeingCancelled(ex, linkedTokenSource.Token))
                    {
                        // Avoid rethrowing cancellation that we expect but the caller does not
                    }
```

Reply:
i'm a little confused.  isn't that waht my code is donig? can talk offline as well if that helps :)

Reply:
got it.